src/conn/conn_ckpt.c

```diff
@@ -83,6 +83,14 @@ __ckpt_server(void *arg)
 
 	while (F_ISSET(conn, WT_CONN_SERVER_RUN) &&
 	    F_ISSET(conn, WT_CONN_SERVER_CHECKPOINT)) {
+		/*
+		 * Wait...
+		 * NOTE: If the user only configured logsize, then usecs
+		 * will be 0 and this wait won't return until signalled.
+		 */
+		WT_ERR(
+		    __wt_cond_wait(session, conn->ckpt_cond, conn->ckpt_usecs));
+
```

Comment:
It may not be obvious why this fixes anything, but in case `ckpt_usecs == 0` (wait forever), we'll always re-check the connection before going to sleep.  The close code carefully clears those flags *then* signals, so with this change we'll see one or the other and won't get stuck.